File: android/lib/src/main/kotlin/com/noise/protocol/state/CipherState.kt
package com.noise.protocol.state

import com.noise.protocol.NoiseException
import com.noise.protocol.crypto.CipherSuite

/**
 * Holds a cipher key and nonce counter for AEAD encryption/decryption.
 */
class CipherState(internal val suite: CipherSuite = CipherSuite.NOISE_25519_CHACHAPOLY_SHA256) {
    private var k: ByteArray? = null
    private var n: Long = 0

    fun initializeKey(key: ByteArray?) {
        k = key
        n = 0
    }

    val hasKey: Boolean get() = k != null

    fun setNonce(nonce: Long) { n = nonce }
    fun getNonce(): Long = n

    /** Encrypt with associated data. If no key is set, returns plaintext. */
    fun encryptWithAd(ad: ByteArray, plaintext: ByteArray): ByteArray {
        val key = k ?: return plaintext
        if (n >= Long.MAX_VALUE - 1) throw NoiseException.NonceExhausted()
        val ct = suite.encrypt(key, n, ad, plaintext)
        n++
        return ct
    }

    /** Decrypt with associated data. On failure, nonce is NOT incremented. */
    fun decryptWithAd(ad: ByteArray, ciphertext: ByteArray): ByteArray {
        val key = k ?: return ciphertext
        val pt = try {
            suite.decrypt(key, n, ad, ciphertext)
        } catch (e: Exception) {
            throw NoiseException.DecryptionFailed()
        }
        n++
        return pt
    }

    /** Derives a new cipher key from the current one (one-way). */
    fun rekey() {
        val key = k ?: throw NoiseException.NoKey()
        val zeros = ByteArray(32)
        val newKey = suite.encrypt(key, Long.MAX_VALUE, ByteArray(0), zeros)
        k = newKey.copyOfRange(0, 32)
    }
}

File: ios/Sources/NoiseProtocol/State/CipherState.swift
import Foundation

/// Holds a cipher key and nonce counter for AEAD encryption/decryption.
public final class CipherState: @unchecked Sendable {
    internal let suite: CipherSuite
    private var k: Data?
    private var n: UInt64 = 0

    public init(suite: CipherSuite = .noise_25519_ChaChaPoly_SHA256) {
        self.suite = suite
    }

    public func initializeKey(_ key: Data?) {
        k = key
        n = 0
    }

    public var hasKey: Bool { k != nil }

    public func setNonce(_ nonce: UInt64) {
        n = nonce
    }

    public func getNonce() -> UInt64 { n }

    /// Encrypt with associated data. If no key is set, returns plaintext.
    public func encryptWithAd(_ ad: Data, plaintext: Data) throws -> Data {
        guard let key = k else { return plaintext }
        guard n < UInt64.max - 1 else { throw NoiseError.nonceExhausted }
        let ct = try suite.encrypt(key, n, ad, plaintext)
        n += 1
        return ct
    }

    /// Decrypt with associated data. On failure, nonce is NOT incremented.
    public func decryptWithAd(_ ad: Data, ciphertext: Data) throws -> Data {
        guard let key = k else { return ciphertext }
        let pt: Data
        do {
            pt = try suite.decrypt(key, n, ad, ciphertext)
        } catch {
            throw NoiseError.decryptionFailed
        }
        n += 1
        return pt
    }

    /// Derives a new cipher key from the current one (one-way).
    public func rekey() throws {
        guard let key = k else { throw NoiseError.noKey }
        let zeros = Data(repeating: 0, count: 32)
        let newKeyData = try suite.encrypt(key, UInt64.max, Data(), zeros).prefix(32)
        k = Data(newKeyData)
    }
}
